test/typecheck/main.go

```diff
@@ -0,0 +1,310 @@
+	a.donePaths[dir] = nil
+
+	// Create the AST by parsing src.
+	fs, err := parser.ParseDir(a.fset, dir, nil, parser.AllErrors)
+
+	if err != nil {
+		fmt.Println("ERROR(syntax)", err)
+		a.failed = true
+		return
+	}
+
+	if len(fs) > 1 && *verbose {
+		fmt.Println("multiple packages in dir:", dir)
+	}
+
+	for _, p := range fs {
+		// returns first error, but a.handleError deals with it
+		if *verbose {
+			fmt.Printf("path: %s package: %s\n", dir, p.Name)
+		}
+		a.typeCheck(dir, a.filterFiles(p.Files))
+	}
+}
+
+// filterFiles restricts a list of files to only those that should be built by
+// the current platform. This includes both build suffixes (_windows.go) and build
+// tags ("// +build !linux" at the beginning).
+func (a *analyzer) filterFiles(fs map[string]*ast.File) []*ast.File {
+	files := []*ast.File{}
+	for _, f := range fs {
+		fpath := a.fset.File(f.Pos()).Name()
+		dir, name := filepath.Split(fpath)
+		matches, err := a.ctx.MatchFile(dir, name)
+		if err != nil {
+			panic(err)
+		}
+		if matches {
+			files = append(files, f)
+		}
+	}
+	return files
+}
+
+func (a *analyzer) typeCheck(dir string, files []*ast.File) error {
+	info := types.Info{
+		Defs: make(map[*ast.Ident]types.Object),
+		Uses: make(map[*ast.Ident]types.Object),
+	}
+
+	// NOTE: this type check does a *recursive* import, but srcimporter
+	// doesn't do a full type check (ignores function bodies)-- this has
+	// some additional overhead.
+	//
+	// This means that we need to ensure that typeCheck runs on all
+	// code we will be compiling.
+	//
+	// TODO(rmmh): Customize our forked srcimporter to do this better.
+	sema, err := a.conf.Check(dir, a.fset, files, &info)
+	if err != nil {
+		return err // type error
+	}
+
+	// A significant fraction of vendored code only compiles on Linux,
+	// but it's only imported by code that has build-guards for Linux.
+	// Track vendored code to type-check it in a second pass.
+	// fileDir := filepath.Dir(a.fset.Position(files[0].Pos()).Filename)
+	for _, p := range sema.Imports() {
+		// p, err := a.conf.Importer.(types.ImporterFrom).ImportFrom(path, dir, mode)
+		if strings.HasPrefix(p.Path(), "k8s.io/kubernetes/vendor/") {
+			vendorPath := p.Path()[len("k8s.io/kubernetes/"):]
+			if *verbose {
+				fmt.Println("recursively checking vendor path:", vendorPath)
+			}
+			a.collect(vendorPath)
+		}
+	}
+
+	if *defuses {
+		for id, obj := range info.Defs {
+			fmt.Printf("%s: %q defines %v\n",
+				a.fset.Position(id.Pos()), id.Name, obj)
+		}
+		for id, obj := range info.Uses {
+			fmt.Printf("%s: %q uses %v\n",
+				a.fset.Position(id.Pos()), id.Name, obj)
+		}
+	}
+
+	return nil
+}
+
+type collector struct {
+	dirs []string
+}
+
+// handlePath walks the filesystem recursively, collecting directories,
+// ignoring third_party and staging directories.
+func (c *collector) handlePath(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+	if path == ".git" ||
+		strings.Contains(path, "staging") ||
```

Comment:
why contains and not ==?

Reply:
Also why don't we want to check code in staging/src/k8s.io?

Reply:
The staging code doesn't type check successfully -- not totally sure why. It's a duplicate of code in the repo, so *theoretically* we should already have coverage, right?

Reply:
We check vendor/ recursively, including the symlinked directories. Checking staging directly doesn't work.